Fix unit tests for core.internal.convert and rt.util.typeinfo

jpf91 opened the pull request:
https://bugzilla.gdcproject.org/show_bug.cgi?id=309

@ibuclaw Here for review an CI. Once you think this is ready, I'll post to gcc-patches.

jpf91 committed:
Fix IdentityExp comparison for complex floats

https://bugzilla.gdcproject.org/show_bug.cgi?id=309

ibuclaw commented:
What if `e1` or `e2` have a side effect? They should be a save_expr in that case.

ibuclaw commented:
Actually, to clarify, they should be saved before extracting the real and imag parts.

jpf91 commented:
So can  I just move `d_save_expr` out of `build_float_identity` and apply it on the 'top-level' expressions?

jpf91 commented:
OK.

ibuclaw commented:
> So can I just move d_save_expr out of build_float_identity and apply it on the 'top-level' expressions?

Right.  You don't need to save anything for ordinary float comparison.

jpf91 commented:
> You don't need to save anything for ordinary float comparison (I might be forgetting something though).

The original code did save float expressions though, so I've also kept that in the commit I just pushed.

jpf91 commented:
Is this ready for gcc-patches?

ibuclaw commented:
I don't see why not. You could always just post to gcc-patches anyway and discuss the patch and make changes there. ;-)

I am yet to get round to fixing up this repo to use as a mirror/ci.

jpf91 committed:
Fix hashing of complex reals

by backporting https://github.com/dlang/druntime/pull/2356

From druntime commit 29ce0543cb62229f005b2bc8540416dbccd1130e

jpf91 committed:
Enable tests for rt.util.typeinfo and core.internal.convert

jpf91 commented:
@ibuclaw 
* Patch for first commit posted here: https://gcc.gnu.org/ml/gcc-patches/2018-11/msg02019.html
* Backport of https://github.com/dlang/druntime/pull/2356 in second commit, for CI testing
* Third commit removes special cases from Makefiles

Note that the `XGDC` bugs in `convert.d` seem to be fixed as well, or at least I can't reproduce these on x86_64.
